cpu/rpx0xx/periph/uart.c

```diff
@@ -0,0 +1,215 @@
+
+    io_reg_atomic_set(&(dev->UARTLCR_H.reg), 0);
+}
+
+int uart_mode(uart_t uart, uart_data_bits_t data_bits, uart_parity_t uart_parity,
+              uart_stop_bits_t stop_bits)
+{
+    UART0_Type *dev = uart_config[uart].dev;
+
+    uint32_t config = (data_bits << UART0_UARTLCR_H_WLEN_Pos);
+
+    switch (uart_parity) {
+    case UART_PARITY_NONE:
+        break;
+    case UART_PARITY_EVEN:
+        config |= UART0_UARTLCR_H_EPS_Msk | UART0_UARTLCR_H_PEN_Msk;
+        break;
+    case UART_PARITY_ODD:
+        config |= UART0_UARTLCR_H_PEN_Msk;
+        break;
+    default:
+        return UART_NOMODE;
+    }
+
+    if (stop_bits == UART_STOP_BITS_2) {
+        config |= UART0_UARTLCR_H_STP2_Msk;
+    }
+
+    dev->UARTLCR_H.reg = config;
+
+    return UART_OK;
+}
+
+void uart_init_pins(uart_t uart)
+{
+    gpio_init(uart_config[uart].tx_pin, GPIO_OUT);
+    gpio_set_function_select(uart_config[uart].tx_pin, FUNCTION_SELECT_UART);
+    if (ctx[uart].rx_cb) {
+        gpio_init(uart_config[uart].rx_pin, GPIO_IN_PU);
+        gpio_set_function_select(uart_config[uart].rx_pin, FUNCTION_SELECT_UART);
+    }
+}
+
+void uart_deinit_pins(uart_t uart)
+{
+    *gpio_io_register(uart_config[uart].tx_pin) = 0;
+    *gpio_pad_register(uart_config[uart].tx_pin) = 0;
+    SIO->GPIO_OE_CLR.reg = 1LU << uart_config[uart].tx_pin;
+    if (ctx[uart].rx_cb) {
+        *gpio_io_register(uart_config[uart].rx_pin) = 0;
+        *gpio_pad_register(uart_config[uart].rx_pin) = 0;
+    }
+}
+
+void uart_poweron(uart_t uart)
+{
+    uint32_t reset_bit_mask = (uart) ? RESETS_RESET_uart1_Msk : RESETS_RESET_uart0_Msk;
+    periph_reset(reset_bit_mask);
+    periph_reset_done(reset_bit_mask);
+}
+
+void uart_poweroff(uart_t uart)
+{
+    uart_deinit_pins(uart);
+    periph_reset((uart) ? RESETS_RESET_uart1_Msk : RESETS_RESET_uart0_Msk);
+}
+
+int uart_init(uart_t uart, uint32_t baud, uart_rx_cb_t rx_cb, void *arg)
+{
+    if (uart >= UART_NUMOF) {
+        return UART_NODEV;
+    }
+
+    UART0_Type *dev = uart_config[uart].dev;
+    ctx[uart].rx_cb = rx_cb;
+    ctx[uart].arg = arg;
+    {
+        static const mutex_t locked = MUTEX_INIT_LOCKED;
+        tx_locks[uart] = locked;
+    }
+
+    uart_poweron(uart);
+    _set_symbolrate(uart, baud);
+
+    if (uart_mode(uart, UART_DATA_BITS_8, UART_PARITY_NONE, UART_STOP_BITS_1) != UART_OK) {
+        return UART_NOMODE;
+    }
+
+    /* enable UART */
+    uint32_t enable_mask = UART0_UARTCR_UARTEN_Msk | UART0_UARTCR_TXE_Msk;
+    if (rx_cb != NULL) {
+        enable_mask |= UART0_UARTCR_RXE_Msk;
+    }
+    io_reg_atomic_set(&dev->UARTCR.reg, enable_mask);
+    uart_init_pins(uart);
+
+    _irq_enable(uart);
+
+    return UART_OK;
+}
+
+void uart_write(uart_t uart, const uint8_t *data, size_t len)
+{
+    UART0_Type *dev = uart_config[uart].dev;
+
+    for (size_t i = 0; i < len; i++) {
+        dev->UARTDR.reg = data[i];
+        mutex_lock(&tx_locks[uart]);
```

Comment:
The mutex lock leads to a deadlock for me if I try to `printf()` within an interrupt routine.
For example we could instead block if the UART Tx FIFO is full:
```C
while (dev->UARTFR.bit.TXFF) {}
```

Reply:
The FIFO is not used by the driver. It would have to busy wait until the char is send.

IMO we really should just disallow using UART during IRQ. We do so for every other periph, I don't know why UART is special in this regard.

Reply:
Fixed